File: mod-common/src/main/java/com/dreamdisplays/net/PremiumPacket.java
package com.dreamdisplays.net;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;
import com.dreamdisplays.PlatformlessInitializer;

// Packet for indicating premium status
public record PremiumPacket(boolean premium) implements CustomPacketPayload {
    public static final Type<PremiumPacket> PACKET_ID =
            new Type<>(Identifier.fromNamespaceAndPath(PlatformlessInitializer.MOD_ID, "premium"));

    public static final StreamCodec<FriendlyByteBuf, PremiumPacket> PACKET_CODEC =
            StreamCodec.of(
                    (buf, packet) -> ByteBufCodecs.BOOL.encode(buf, packet.premium()),
                    (buf) -> {
                        boolean premium = ByteBufCodecs.BOOL.decode(buf);
                        return new PremiumPacket(premium);
                    });

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return PACKET_ID;
    }
}

File: common/src/main/kotlin/com/dreamdisplays/net/s2c/PremiumPacket.kt
package com.dreamdisplays.net.s2c

import com.dreamdisplays.net.common.helpers.createType
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamCodec.of
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.network.protocol.common.custom.CustomPacketPayload.Type

data class PremiumPacket(val premium: Boolean) : CustomPacketPayload {
    override fun type(): Type<out CustomPacketPayload> = PACKET_ID

    companion object {
        val PACKET_ID: Type<PremiumPacket> = createType("premium")
        val PACKET_CODEC: StreamCodec<FriendlyByteBuf, PremiumPacket> = of(
            { buf, packet -> buf.writeBoolean(packet.premium) },
            { buf -> PremiumPacket(buf.readBoolean()) }
        )
    }
}
